SSO Overhaul, Authentication Extenders

askvortsov1 opened the pull request:
Fixes flarum/framework#2043 
Fixes Part of flarum/framework#1891 
Fixes flarum/issue-archive#380 
Fixes flarum/issue-archive#236 
Fixes flarum/issue-archive#332
Fixes flarum/issue-archive#223 
Fixes flarum/issue-archive#285 
- probably more

**Changes proposed in this pull request:**
- [x] Deprecate disorganized components of old authentication system
- [x] Centralize system for extensions to register new authentication systems
- [x] Add extender to add an sso driver
- [x] Sso registration respects disabled signups.
- [x] Add new admin pane, allowing forum admins to:
   - [x] Enable/disable authentication providers
   - [x] Enable/disable signups (move from permissions pane)
   - [x] Enable/disable username/password registration
- [x] Add new user settings page, allowing users to link and unlink external auth providers from their account
   - [x] Don't allow users to unlink sso providers if username/password authentication is disabled, and the account in question only has 1 linked provider.
- [x] Add a console command to re-enable username/password login (if SSO breaks we don't want to strand forum admin).
- [x] Add 'trust emails' feature. If a user is not logged in, and SSO returns the email of an existing user, an error message will be shown if this SSO Driver doesn't have 'trust emails' enabled in the admin pane. If trust emails is enabled for this driver, the SSO provider will be linked to the user's account.
- [x] Automatically display login buttons based off of meta method of driver.
- [x] Rewrite bundled extensions to use this new system
   - [x] Auth Github
   - [x] Auth Facebook
   - [x] Auth Twitter
- [x] Add integration tests for SSO extender methods/functionality test (relies on flarum/framework#2052)
- [x] Extensive manual tests

TO ADD AFTER STABLE
- Ability to test SSO providers in admin panel before enabling (shouldn't be too difficult)
- Add (non-accidential) support for linking multiple identifiers from a provider to a flarum user.
- Manage button meta appearance from auth settings pane
- Manage provider settings through auth settings pane?
- Add extender method, full support for multifactor authentication.

**Reviewers should focus on:**
- How should we standardize nomenclature? IE sso vs auth vs external auth? driver vs provider? Right now, it's kind of a mess of all these.
- Does the sso flow described in SsoController accurately describe what we want? If not, what should we change/improve?
- Any things I missed?

**Screencasts**
Disable username/password auth:
https://youtu.be/kN4XT8BwUIQ

Disabling all sso providers force enables user/pass auth
https://youtu.be/OvS1jxHj5QQ

Disabling signup:
https://youtu.be/TnwrDFgCqnM

Auth Panel controls buttons, Registration works
https://youtu.be/EQchPk7oQuc

Linking/unlinking accounts works
https://youtu.be/teHYWFbTzQY

Can't unlink last provider if username/password auth disabled
https://youtu.be/YcMkBNWORgY

SSO registration respects disabled signups:
https://youtu.be/UCYJWyJB07I

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).

Taraflex committed:
Protect dismissible modals from closing by ESC key

luceos committed:
Notifications into the queue

Forces notifications into a dedicated SendNotificationsJob and passed
to the queue.

- One static method re-used in the job ::getAttributes, is that okay or
  use a trait?
- Do we want to use this solution and refactor into a better Hub after
  stable, postpone this implementation or use it in b11?

luceos committed:
Moved sending emails to the syncer

This separates sending each individual mail, thus hardening the app.
There are still many improvements possible in this code, e.g. chaining
these commands, making emails just another notification type and
listening to the Notify event instead. We can postpone this to a later
stable release.

franzliedke committed:
Install prettier

askvortsov1 committed:
Deprecated old sso systems

askvortsov1 committed:
Added sso driver

askvortsov1 committed:
Added extendable singleton of drivers to forum service provider.

askvortsov1 committed:
Added SsoResponse class, based off of Registration, but tweaked to fit new system

askvortsov1 committed:
Added SSO Controller

askvortsov1 committed:
Added first version of auth extender, which supports adding drivers

askvortsov1 committed:
Bugfixes

askvortsov1 committed:
Add abstract auth lifecycle handler to auth extender

askvortsov1 committed:
Added new admin site pane with options for enabling/disabling signups (moved from permissions page), enabling/disabling username password authentication

askvortsov1 committed:
Got auth pane working to manage whether SSO providers are enabled

askvortsov1 committed:
Remove reset password button from settings page if username password authentication is disabled

askvortsov1 committed:
Uncommented out driver enabled check (now that implemented in admin), fixed trust emails check to use settings not driver method

askvortsov1 committed:
Added console command to reenable username/password authentication

askvortsov1 committed:
Fixed incorrect translator key for trust emails setting

askvortsov1 committed:
Added SSO linking.unlinking grid

askvortsov1 committed:
Added sso linking/unlinking grid to settings page

askvortsov1 committed:
Manage user pass auth fields appearing in body(), not in fields() for login and signup modals

askvortsov1 committed:
Added delete login provider controller

askvortsov1 committed:
Added serialization of enabled sso drivers to forum attributes

askvortsov1 committed:
Added serialization of user's linked sso drivers to user attributes

askvortsov1 committed:
Automatically generate login buttons based off of enabled LDAP providers

askvortsov1 committed:
Display sso driver slug if meta name not provided in login button

askvortsov1 committed:
Refactored SsoController logic flow

askvortsov1 committed:
Added support for testSuccess response to authenticationComplete handler

tankerkiller125 commented:
> How should we standardize nomenclature? IE sso vs auth vs external auth? driver vs provider? Right now, it's kind of a mess of all these.

The nomenclature is pretty simple for me at least. 

SSO is Representative of the whole sign in experience through 3rd party companies/services

Auth to me at least describes any way of logging in to the system (AKA are you authorized to access this resource)

External auth, I think is something we can get rid of, either that or we need to replace SSO with it depending on if we want to be technically accurate or we want it to be user friendly (SSO being technically accurate)

When it comes to driver vs provider the way I see it is that the driver is the implementation between flarum and the 3rd party company/service. Provider to me at least means the 3rd party company/service itself.

> Does the sso flow described in SsoController accurately describe what we want? If not, what should we change/improve?

The flow looks correct to me, but I would have to test it myself to get a full understanding of how it works.

> Should the "disable username/password auth" slider be disabled if no SSO providers are enabled? I'm also leaning in this direction.

Yes, I'd rather not let people break their site, even if admins can bypass it (which it appears they can?) I just think it would create a support headache if we don't disable the slider if there are no SSO provider enabled, further I think that the slider should automatically slide to be turned off if the user disables all of the 3rd party providers.

Also before you go and update all the bundled 3rd party providers to in theory support this system I would recommend you wait until a decision is made on flarum/framework#2006

askvortsov1 committed:
Move logic that provides supported drivers to AdminPayload instead of as an API controller

askvortsov1 committed:
Don't use for of loop, which isn't supported in IE

askvortsov1 committed:
Don't allow linking multiple identifiers from a provider with a user (might add support after stable).

askvortsov1 committed:
Disable switching the enable_user_pass_auth checkbox if all SSO providers are disabled, automatically set it to "on" if all SSO providers are disabled

askvortsov1 committed:
Fixed disabled switch checkbox CSS

askvortsov1 commented:
>SSO is Representative of the whole sign in experience through 3rd party companies/services
>
>Auth to me at least describes any way of logging in to the system (AKA are you authorized to access this resource)
>
>External auth, I think is something we can get rid of, either that or we need to replace SSO with it depending on if we want to be technically accurate or we want it to be user friendly (SSO being technically accurate)
>
>When it comes to driver vs provider the way I see it is that the driver is the implementation between flarum and the 3rd party company/service. Provider to me at least means the 3rd party company/service itself.

I agree with this interpretation, I'll standardize this. I'm also going to rename LoginProvider (and the associated table) to SsoProvider, as that more accurately describes what it does (username password login doesn't use LoginProviders afaik.

>Yes, I'd rather not let people break their site, even if admins can bypass it (which it appears they can?) I just think it would create a support headache if we don't disable the slider if there are no SSO provider enabled, further I think that the slider should automatically slide to be turned off if the user disables all of the 3rd party providers.

Done.

> Also before you go and update all the bundled 3rd party providers to in theory support this system I would recommend you wait until a decision is made on flarum/framework#2006

One of my main reasons to do this is to give the core dev team an easier way to test the new functionality when reviewing this PR. Besides, even if the extensions are getting moved to FoF, might as well make their lives easier too :D

askvortsov1 committed:
Fixed style attribute on sso login buttons

askvortsov1 committed:
Fixed disable checkbox switch

askvortsov1 committed:
Don't allow unlinking a login provider if username/password authentication is disabled, and it is the only login provider

askvortsov1 committed:
Rename linkedProviders method on user to linkedSsoProviders (no need to deprecate since new in this PR)

askvortsov1 committed:
Rename login providers to sso providers, as that's more accurate

askvortsov1 committed:
Renamed linkedSsoProviders to ssoProviderNames

askvortsov1 committed:
Made sure all internal uses of loginProviders use ssoProviders instead.

askvortsov1 committed:
Deprecated provideTrustedEmail in favor of provideEmail, as email trust is now determined in admin settings

askvortsov1 committed:
Standardized naming

askvortsov1 committed:
Started integration tests

askvortsov1 committed:
Fix $user-isAdmin to $actor->isAdmin when assessing whether testing sso

askvortsov1 committed:
Fixed user being passed to linkProvider instead of actor when linking new account to already logged in user

askvortsov1 committed:
Set SsoResponse payload as array by default

askvortsov1 committed:
Finished auth sso extender/functionality integration test, renamed to AuthSsoTest

askvortsov1 committed:
AuthSsoTest: save settings in variable to avoid regenerating with every request.

askvortsov1 committed:
Fix missing 'return $this' on auth extender authLifecycleHandler method

askvortsov1 committed:
Fix incorrect argument passed in abstract auth lifecycle handler

askvortsov1 committed:
Added auth lifecycle extender test

askvortsov1 committed:
Fixed removing ssoProviders UX, removed leftover console messages

askvortsov1 committed:
Fixed login buttons not being displayed (after new naming)

askvortsov1 committed:
Sso registation should respect disabled signups

askvortsov1 committed:
Move registering response fragment out into constant

clarkwinkelmann reviewed (changes requested):
I read through the code and here are some comments I have.

Also I noticed some files don't have the ending newline. Not sure if our automated tools will fix that formatting or not.

askvortsov1 committed:
Remove lifecycle handler component from auth extender

askvortsov1 committed:
Rename 'user pass auth' to 'password auth

askvortsov1 committed:
Remove console command to enable password authentication

askvortsov1 committed:
Changed uses of ==  where compared to a primitive to ===

askvortsov1 committed:
Automatically enable password authentication and save if opening panel and its disabled while all sso drivers are disabled. Also, fixed switch state not working with both '1' and true while being off with '0' and 'false

askvortsov1 committed:
Use object for login button styles instead of string

askvortsov1 committed:
Removed unneeded @important in less

askvortsov1 committed:
Removed unnecessary loading on synchronous auth page load

askvortsov1 committed:
Filter sso drivers for enabled before making them

askvortsov1 committed:
Dont instantiate components when not necessary

KyrneDev committed:
Change default discussion comment count

This allows new public discussions to be immediately visible by users.

hasan-ozbey committed:
fix modals

askvortsov1 committed:
When creating access tokens, assume username password authentication is on by default

askvortsov1 committed:
Fixed Auth SSO tests to work with BuildsHttpRequest

hasan-ozbey committed:
Update Modal.less

hasan-ozbey committed:
Update ModalManager.js

luceos committed:
Resolved enabled extension test

The getEnabled method returns all extensions (previously) enabled, yet manually
uninstalled through composer. This does not reference the exact, current state
of the forum. getEnabledExtensions returns a list where the getEnabled list
is filtered on the extensions found in the composer installed.json file.

tankerkiller125 committed:
Fix test and extender for middleware (#2084)

franzliedke committed:
Convert more controller tests to request tests

franzliedke committed:
Rename API tests for more consistency

I could not come up with a noun for the new "UpdateTest" for users, so
this is easier in terms of consistency.

luceos committed:
fixes checking for enabled extension and correct pointer of https://github.com/flarum/core/commit/30c6ea9912743cc0b4568c4352fca629505ef87b

franzliedke committed:
Convert last two controller tests to request tests

franzliedke committed:
Remove Notifying event for now

As discussed with @luceos, let's add this once the use case comes up. It
might be a left-over from a previous state of this PR anyway.

franzliedke committed:
Remove unnecessary constructor parameter

franzliedke committed:
Add BC layer for notification blueprints

This gives extension authors time to add the new `getAttributes()`
method to their `BlueprintInterface` implementations.

The layer itself is easy to remove in beta.14.

franzliedke committed:
Add new method to DiscussionRenamedBlueprint

franzliedke committed:
Merge pull request #1931 from flarum/dk/1869-queue-notifications

Notifications into the queue

franzliedke committed:
Configure prettier via JSON file

franzliedke committed:
Add NPM shortcut for running prettier

franzliedke committed:
Make tests compatible with PHPUnit 8

franzliedke committed:
Laravel: Stop calling deprecated fire() method

This has been deprecated and removed from the contract for a long time,
and it will be completely dropped in v5.8, our next upgrade target.

franzliedke committed:
Use UrlGenerator over Application for base URL

We need to get rid of this god class, as Laravel's Application contract
gets even bigger with 5.8. To avoid having to add all these methods, we
should try to stop using it where we can.

askvortsov1 committed:
Send emails through the queue

flarum-bot committed:
Bundled output for commit 6bbd603a4126e5f373aaa9ad663d6528531cbb38 [skip ci]

luceos committed:
fixed an issue with Post--by-start-user for discussions that contain posts of deleted users

flarum-bot committed:
Bundled output for commit 2e2aa8747e26e6de0ee5ca6059b7be8973d3fea2 [skip ci]

hasan-ozbey committed:
fix extensions page layout

franzliedke committed:
Revert BC breaks around notification blueprints

No need for breaking backwards compatibility here - encapsulating the
logic for `getAttributes()` in one place turns out to be quite useful.

Refs #1931.

franzliedke committed:
Deprecations: Add removal dates and replacements

franzliedke committed:
Update JS dependencies to secure versions

flarum-bot committed:
Bundled output for commit dd43e49d0afc49676c88b6b6215f6b7112456825 [skip ci]

franzliedke committed:
Add integration test for API root endpoint

franzliedke committed:
ForumSerializer: Use UrlGenerator for base URLs

The test from the previous commit proves this works as intended. :)

This is one more step in trying to avoid the widespread usage of the
`Application` godclass.

Refs #2055.

franzliedke committed:
Simplify uploads, inject filesystem instances

This avoids injecting the Application god class and assembling default
file locations in multiple places.

In addition, we no longer use the `MountManager` for these uploads. It
only added complexity (by moving tmp files around) and will not be
available in the next major release of Flysystem.

Note: Passing PSR upload streams to Intervention Image requires an
explicit upgrade of the library. (Very likely, users have already
updated to the newer versions, as the old constraint allowed it, but
we should be explicit for correctness' sake.)

askvortsov1 committed:
Add console extender (#2057)

* Made the console command system extender-friendly

* Added console extender

* Added ConsoleTestCase to integration tests

* Added integration tests for console extender

* Marked event-based console extension system as deprecated

* Moved trimming command output of whitespace into superclass

* Renamed 'add' to 'command'

* Added special processing for laravel commands

* Code style fixes

* More style fixes

* Fixed $this->container

SychO9 committed:
Fix new post injected above unread sticky (#1868)

Refresh the discussion list instead of prepending the new post

flarum-bot committed:
Bundled output for commit 30942bdf38f665ec771a8c1fc06d83aca7e2220f [skip ci]

askvortsov1 committed:
Added CSRF Extender (#2095)

askvortsov1 committed:
StyleCI fixes

askvortsov1 committed:
More StyleCi fixes

askvortsov1 committed:
Move logic to not add login/signup fields into fields method.

askvortsov1 committed:
Refactored out some m.redraw statements and moved component initialization into the view method

franzliedke committed:
Validate PSR-compatible file upload

Instead of converting the uploaded file object to an UploadedFile
instance from Symfony, because the latter is compatible with
Laravel's validation, let's re-implement the validation for the
three rules we were using.

The benefit: we can now avoid copying the uploaded file to a
temporary location just to do the wrapping.

In the next step, we will remove the temporary file and let the
uploader / Intervention Image handle the PSR stream directly.

franzliedke committed:
Permissions page: Tweak icon styling

- Give them a fixed width (independent of font library)
- Center the icons in their column
- De-emphasize the icons by applying a muted color

Fixes #2016, closes #2018.

flarum-bot committed:
Bundled output for commit fd4c0d30d8af429c7b58008aa1ef15873c81942b [skip ci]

hasan-ozbey committed:
Fix mobile notification bubble on colored header (#2109)

askvortsov1 committed:
Only show login indicator when unlinking (as no way to remove login indicator when linking if sso window closed)

w-4 committed:
Fix PostStreamScrubber click (#1945)

flarum-bot committed:
Bundled output for commit 117c2f65ac54002cd9d592a7e5ee7aebbacfb510 [skip ci]

askvortsov1 committed:
Improved UI of Switch with loading indicator (#2039)

* Moved loading indicator outside of checkboxes to improve ui
* Made loading indicator more visible, fade label when switch is loading

askvortsov1 committed:
Add event extender (used for domain events) (#2097)

askvortsov1 committed:
Mail Extender (#2012)

This allows registering new drivers, or overwriting existing ones.

askvortsov1 committed:
Add year, localization support for displaying things older than 1 year (#2034)

flarum-bot committed:
Bundled output for commit dbfae0b55e69c49a0c1033c73c4e58522687746a [skip ci]

franzliedke committed:
Tweak mail extender (tests)

- Use private over protected
- Use "public" API for building requests in tests
- Add more assertions
- Formatting
- Use correct parameter order for assertions

Refs #2012.

franzliedke committed:
Tweak event extender (tests)

- Inject contract, not implementation
- Do not dispatch event in test, let the core do that
- Ensure the relevant database tables are reset prior to the test
- Use correct parameter order for assertions

Refs #2097.

franzliedke committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

franzliedke committed:
Merge pull request #2117 from flarum/fl/2055-streamline-uploads

Simplify uploads, avoid Application contract

franzliedke committed:
Setup husky for automatic formatting before commit

franzliedke committed:
CI: Run prettier to check for JS code formatting

franzliedke committed:
Merge pull request #2099 from flarum/fl/prettier

Install prettier for consistent JS styling

franzliedke committed:
Run prettier for all JS files

flarum-bot committed:
Bundled output for commit 89ef14faf13e4d661ee9f4f636a32578c73bf760 [skip ci]

clarkwinkelmann committed:
Make AbstractPolicy compatible with both object and class as $model (#1977)

franzliedke committed:
Inline test class

Refs #1977.

rob006 committed:
Load only translations for enabled extensions from language packs (#2020)

fix #1837

Co-authored-by: Daniel Klabbers <daniel+git@klabbers.email>

franzliedke committed:
Format code

- Early returns
- Comments
- Write variables only when needed

Refs #2020.

franzliedke committed:
Add compatiblity with Composer 2.0

- The structure of vendor/composer/installed.json will change.
- The same file will now contain the relative path to package locations.

References:
- https://github.com/composer/composer/blob/master/UPGRADE-2.0.md
- https://php.watch/articles/composer-2

franzliedke committed:
Use Container contract where easily possible

Less usages of the Application god-class simplifies splitting it up.

Refs #2055.

franzliedke committed:
Statically access Flarum version everywhere

One less reason to inject the huge Application class.

Refs #2055.

dsevillamartin reviewed (changes requested):
This PR seems to have commits that shouldn't be here. Not sure what happened - perhaps you rebased your branch and the existing commits have different SHAs so they are new?

267 files cannot be changed by this PR 🙂

tankerkiller125 reviewed (changes requested):
My original review was a case of not paying close enough attention to which PR I was on
